first draft checklists (#30)
fornax-navo/fornax-documentation: notebook_review_process.md

For authors: consider these checklists requirements for your code.

## Who should participate in these reviews?
- JK's suggestion is that if everyone on Fornax helps to write the below checklists, then the whole team does not need to be involved in code reviews.  One developer per tech review, one scientist per science review, from any team.
  

## Science Review Checklist
- Is there a use case in the introduction which motivates the code?  will our community understand this motivation/code?
- Does the code do what the intro says it is going to do?
- Is it scientifically accurate?
- Does it include all three archives HEASARC, MAST, IRSA?\
      - if not, is that justified
- Does it include work linked to a buzzword:
	- big data, spectroscopy, time domain, forced photometry, cloud
- Has each NASA archive been given option to comment on modules for their relevant data access?\
	- Is archival data accessed in the most efficient way according to that archive?

> ## Tech Review Checklist
> - Is every function documented?
> - Does it follow the style guide? https://github.com/spacetelescope/style-guides/blob/master/guides/jupyter-notebooks.md
> - Is everything that gets done more than once turned into a function?
> - Are errors handled appropriately?
> - Are most lines of code commented?
> - Is data accessed from the cloud where possible?
> - Is the code parallelized where possible?
> - If the notebook is intended to be scaled up, does it do that efficiently?
> - Is memory usage optimized where possible? 
> - Does the code give warnings that need to be followed up?
> - Is the requirements.txt file complete and not include un-used libraries?

Reviewer: ```suggestion
## Tech Review Checklist
- Documentation:
    - Does the notebook and accompanying code follow the [style guide](https://github.com/spacetelescope/style-guides/blob/master/guides/jupyter-notebooks.md)?
    - Do all code cells have corresponding narratives/comments?
    - Is every function documented?
- Notebook execution, error handling, etc.:
    - Does the notebook run end-to-end, out of the box?
    - Are errors handled appropriately, with `try`/`except` statements that are narrow in scope?
    - Have warnings been dealt with appropriately, preferably by updating the code to avoid them (i.e., not by simply silencing them)?
- Efficiency:
    - Is data accessed from the cloud where possible?
    - If the notebook is intended to be scaled up, does it do that efficiently?
    - Is the code parallelized where possible?
    - Is memory usage optimized where possible? 
- Cleanup:
    - Have blocks of code that need to be re-used been turned into functions (rather than being duplicated)?
    - Has un-used code been removed (e.g., unused functions and commented-out lines)? 
    - Have un-used libraries been removed from the requirements.txt file and the `import` statements?
```

For authors: consider these checklists requirements for your code.

## Who should participate in these reviews?
- JK's suggestion is that if everyone on Fornax helps to write the below checklists, then the whole team does not need to be involved in code reviews.  One developer per tech review, one scientist per science review, from any team.
  

## Science Review Checklist
- Is there a use case in the introduction which motivates the code?  will our community understand this motivation/code?
- Does the code do what the intro says it is going to do?
- Is it scientifically accurate?
- Does it include all three archives HEASARC, MAST, IRSA?\
      - if not, is that justified
- Does it include work linked to a buzzword:
	- big data, spectroscopy, time domain, forced photometry, cloud

## Tech Review Checklist
- Is every function documented?
- Does it follow the style guide? https://github.com/spacetelescope/style-guides/blob/master/guides/jupyter-notebooks.md
- Is everything that gets done more than once turned into a function?
- Are errors handled appropriately?

> - Are most lines of code commented?

Reviewer 2: I would change this to something along the line:

```suggestion
- Have all code cells have corresponding narratives/comments?
```

- Is archival data accessed in the most efficient way according to that archive?
- Is data accessed from the cloud where possible?
- Is the code parallelized where possible?
- If the notebook is intended to be scaled up, does it do that efficiently?
- Is memory usage optimized where possible? 

> - Does the code give warnings that need to be followed up?

Reviewer 2: ```suggestion
- Does the code give warnings that need to be followed up without a a blanket ignore of all warnings?
```

> - Is the requirements.txt file complete and not include un-used libraries?

Reviewer 2: ```suggestion
- Is the requirements.txt file complete and not include un-used libraries?
- Has the notebooks and code directory has any non-used code that is commented out as obsolete or unused? If yes, clean-up those.
```
